ionic-team/stencil
# feat(compiler): export custom types in compiled output

## Pull request checklist

Please check if your PR fulfills the following requirements:

## Pull request type

 

 

Please check the type of change your PR introduces:


## What is the current behavior?


Currently, types for custom events are not exported from the generated `components.d.ts` file

GitHub Issue Number: N/A


## What is the new behavior?


Types for custom events are now re-exported from the `components.d.ts` file. This allows those types to be easily accessed from the root of the type distribution (so long as the `index.ts` file is set to export all definitions from the `components.d.ts` file):
`import { MyCustomType } from '@my-lib/types';`

Or, when using `dist-custom-elements`, these types can now be accessed from the custom element output:
`import { MyCustomType } from '@my-custom-elements-output';`

## Does this introduce a breaking change?



Any project that currently manually maintains a means of creating type definitions for types created as a part of the Stencil project will either simply need to remove that manual step from their development or update it to re-export the types exported from the generated `component.d.ts` file

## Testing



- Unit and e2e tests continue to pass. 
- New tests added for altered function.
- Tested output in starter projects.
- Built and ran Ionic Framework (this is a case where the above migration would need to occur)

**Testing in a starter**
- Add a prop or event to a Stencil component that leverages a custom type (either defined locally in the file or as a separate type file)
- Build the project
- In the generated `component.d.ts` file, verify that the types are import to created component definitions as well as exported from the file to be used in consuming applications

## Other information


This was originally introduced as a feature in Stencil 2.x, but was reverted when a breaking change was identified.

Initial PR (orginally merged as a part of Stencil 2.x): #3612
Reversion of above PR (due to identified breaking change): #3708
